Add a basic version of avifgainmaputil.

maryla-uc opened the pull request:
It has three commands:
- help
- extractgainmap to save the gain map of a file as an image
- printmetadata to print detailed gain map metadata

Since it's written in c++ I chose to follow the same style guide as the c++ tests.
I placed it in a cpp/ subdirectory to be able to add a .clang-format file.

vrabaud commented on CMakeLists.txt:
```diff
@@ -797,6 +797,16 @@ if(AVIF_BUILD_APPS)
     endif()
+
+    if(AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP)
+        enable_language(CXX)
```
Maybe we should merge all the `if`s at the beginning to set C++ parameters.
This would prevent us from making mistakes when setting the standard to a version somewhere and to a different version somewhere else.

maryla-uc replied:
Done by vrabaud

vrabaud commented on apps/cpp/avifgainmaputil.cc:
```diff
@@ -0,0 +1,386 @@
+}
+
+// Writes an image in any of the supported formats based on the file extension.
+avifResult WriteImage(const avifImage* image, std::string output_filename,
```
const std::string& output_filename

vrabaud commented on apps/cpp/avifgainmaputil.cc:
```diff
@@ -0,0 +1,386 @@
+      return AVIF_RESULT_UNKNOWN_ERROR;
+    }
+  } else if (output_format == AVIF_APP_FILE_FORMAT_AVIF) {
+    std::unique_ptr<avifEncoder, decltype(&avifEncoderDestroy)> encoder(
```
maybe wait for https://github.com/AOMediaCodec/libavif/pull/1733 to get EncoderPtr

vrabaud commented on apps/cpp/avifgainmaputil.cc:
```diff
@@ -0,0 +1,386 @@
+
+  // Returns the name of the command that should be used to invoke it on the
+  // command line.
+  virtual std::string GetName() = 0;
```
const function maybe? And the ones below.

vrabaud commented on apps/cpp/avifgainmaputil.cc:
```diff
@@ -0,0 +1,386 @@
+
+MAIN() {
+  std::vector<std::unique_ptr<ProgramCommand>> commands;
+  commands.emplace_back(std::make_unique<HelpCommand>());
```
maybe an initializer list is easier to read?

maryla-uc replied:
I tried that first but std::vector *copies* the content of the initializer list which means all the commands must be copyable.

vrabaud commented on apps/shared/avifutil.h:
```diff
@@ -40,14 +59,18 @@ extern "C" {
 #define INIT_ARGV()
 #endif
-
 #if defined(_WIN32)
```
`#if defined(_WIN32) && !defined(__cplusplus)` ?

maryla-uc commented on apps/shared/avifutil.h:
```diff
@@ -18,6 +18,25 @@ extern "C" {
+            return 1;                                                                                         \
+        }                                                                                                     \
+    }
+#else
```
https://www.youtube.com/watch?v=HXsU53Veupw

maryla-uc commented on apps/cpp/avifgainmaputil.cc:
Done.

maryla-uc commented on apps/cpp/avifgainmaputil.cc:
```diff
@@ -0,0 +1,386 @@
+avifResult WriteImage(const avifImage* image, std::string output_filename,
+                      int quality, int speed) {
+  quality = Clamp(quality, 0, 100);
+  speed = Clamp(quality, 0, 10);
```
Nice catch.

maryla-uc commented on apps/shared/avifutil.h:
```diff
@@ -18,6 +18,25 @@ extern "C" {
+#ifdef __cplusplus
+#define INIT_ARGV()                                                                                           \
+    if (setlocale(LC_ALL, ".UTF8") == NULL) {                                                                 \
+        fprintf(stderr, "setlocale failed\n");                                                                \
```
Actually it's not needed in the c++ version, I removed the call to FREE_ARGV below.

maryla-uc commented:
Lots of changes: I switched from popl to argparse, integrated vrabaud's changes, split the utility into multiple files, and renamed the cpp/ directory to avifgainmaputil/

y-guyon commented on apps/avifgainmaputil/avifgainmaputil.cc:
```diff
@@ -0,0 +1,111 @@
+  int longest_command_size = 0;
+  for (const std::unique_ptr<ProgramCommand>& command : commands) {
+    longest_command_size =
+        std::max(longest_command_size, (int)command->name().size());
```
`static_cast` in C++  
Alternatively `std::max<int>`

y-guyon commented on ext/argparse.cmd:
```diff
@@ -0,0 +1,21 @@
+: # If you're running this on Windows, be sure you've already run this (from your VC2019 install dir):
+: #     "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Auxiliary\Build\vcvars64.bat"
+
+: # libsharpyuv is part of the libwebp repo.
```
Leftover

y-guyon commented on tests/test_cmd_avifgainmaputil.sh:
```diff
@@ -0,0 +1,56 @@
+  "${AVIFGAINMAPUTIL}" extractgainmap "${INPUT_AVIF_GAINMAP}" "${AVIF_OUTPUT}" --quality 50
+  "${AVIFGAINMAPUTIL}" extractgainmap "${INPUT_AVIF_GAINMAP}" "${JPEG_OUTPUT}"
+  "${AVIFGAINMAPUTIL}" extractgainmap --speed 9 "${INPUT_AVIF_GAINMAP}" "${PNG_OUTPUT}"
+
```
Extra _blank line
